### PR TITLE
feat: cmdline functions

### PR DESCRIPTION
**Pull request by Shougo**
I have added the functions for command line completion.

```vim
autocmd CmdlineChanged * echomsg getcmdcompltype()
autocmd CmdlineChanged * echomsg 'screenpos: ' . getcmdscreenpos()
autocmd CmdlineChanged * echomsg 'pos: ' . getcmdpos()
autocmd CmdlineChanged * echomsg 'cmdline: ' . getcmdline()

echo input('foo ', '', 'dir')
```

https://github.com/hrsh7th/nvim-cmp/issues/938

### CODE REVIEW

File: src/nvim/eval/funcs.c [hunk 1]
```diff
@@ -3188,6 +3195,12 @@ static void f_getcmdpos(typval_T *argvars, typval_T *rettv, FunPtr fptr)
   rettv->vval.v_number = get_cmdline_pos() + 1;
 }
 
+/// "getcmdscreenpos()" function
+static void f_getcmdscreenpos(typval_T *argvars, typval_T *rettv, FunPtr fptr)
+{
+  rettv->vval.v_number = get_cmdline_screen_pos() + 1;
```
Comment (bfredl) [inline src/nvim/eval/funcs.c:3201]: what does this mean? how is position just a number? what happens when `ext_cmdline` is set?

Comment (Shougo) [reply src/nvim/eval/funcs.c:3201]: It means screen cursor position in command line.
It is needed for command line popup feature.

I don't test `ext_cmdline`.  But it should work.

**Comment by Shougo** [comment]
`getcmdscreenpos()` is needed for command line popup in `pum.vim`.  It can be used in `nvim-cmp`.

**Comment by Shougo** [comment]
`:call input('piyo >', '', '')` auto completion is here.

スクリーンショット_2022-04-29_15-44-09


I think it is useful.

Note:  The latest `ddc.vim`, `ddc-cmdline` and `pum.vim` are needed.

**Comment by Shougo** [comment]
Another example is here:

スクリーンショット_2022-04-29_17-46-19

**Comment by Shougo** [comment]
I have added the tests.
The next is for Vim8 version.

**Comment by Shougo** [comment]
I have ported it to Vim8.  But `getcmdcompletion()` does not work in Vim8...
I don't know why.

**Comment by Shougo** [comment]
Please wait until Vim patches are merged.

**Comment by Shougo** [comment]
I have applied https://github.com/vim/vim/commit/79d599b8772022af1d657f368c2fc97aa342c0da#commitcomment-73098124 patch.

File: runtime/doc/usr_41.txt [hunk 2]
```diff
@@ -855,8 +855,11 @@ Buffers, windows and the argument list:
 	swapname()		get the swap file path of a buffer
 
 Command line:					*command-line-functions*
+	getcmdcmpltype()	get the current command line completion
```
Comment (zeertzjq) [inline runtime/doc/usr_41.txt:858]:
```suggestion
	getcmdcompltype()	get the type of the current command line
				completion
```

File: src/nvim/eval.lua [hunk 3]
```diff
@@ -152,8 +152,10 @@ return {
     getcharpos={args=1, base=1},
     getcharsearch={},
     getcharstr={args={0, 1}},
+    getcmdcmpltype={},
```
Comment (zeertzjq) [inline src/nvim/eval.lua:155]:
```suggestion
    getcmdcompltype={},
```

File: src/nvim/eval/funcs.c [hunk 4]
```diff
@@ -3175,6 +3175,13 @@ static void f_getcharsearch(typval_T *argvars, typval_T *rettv, FunPtr fptr)
   tv_dict_add_nr(dict, S_LEN("until"), last_csearch_until());
 }
 
+/// "getcmdcmpltype()" function
+static void f_getcmdcmpltype(typval_T *argvars, typval_T *rettv, FunPtr fptr)
```
Comment (zeertzjq) [inline src/nvim/eval/funcs.c:3179]:
```suggestion
/// "getcmdcompltype()" function
static void f_getcmdcompltype(typval_T *argvars, typval_T *rettv, FunPtr fptr)
```

File: src/nvim/testdir/test_cmdline.vim [hunk 5]
```diff
@@ -1212,4 +1212,16 @@ func Test_recalling_cmdline()
   cunmap <Plug>(save-cmdline)
 endfunc
 
+func Check_completion()
+  call assert_equal('let a', getcmdline())
+  call assert_equal(6, getcmdpos())
+  call assert_equal(7, getcmdscreenpos())
+  call assert_equal('var', getcmdcmpltype())
```
Comment (zeertzjq) [inline src/nvim/testdir/test_cmdline.vim:1219]:
```suggestion
  call assert_equal('var', getcmdcompltype())
```

File: runtime/doc/builtin.txt [hunk 6]
```diff
@@ -179,8 +179,12 @@ getcharmod()			Number	modifiers for the last typed character
 getcharpos({expr})		List	position of cursor, mark, etc.
 getcharsearch()			Dict	last character search
 getcharstr([expr])		String	get one character from the user
+getcmdcmpltype()		String	return the current command-line
+					completion
```
Comment (zeertzjq) [inline runtime/doc/builtin.txt:183]:
```suggestion
getcmdcompltype()		String	return the current command-line
					completion
```

File: runtime/doc/builtin.txt [hunk 7]
```diff
@@ -179,8 +179,12 @@ getcharmod()			Number	modifiers for the last typed character
 getcharpos({expr})		List	position of cursor, mark, etc.
 getcharsearch()			Dict	last character search
 getcharstr([expr])		String	get one character from the user
+getcmdcompltype()		String	return the current command-line
+					completion
```
Comment (zeertzjq) [inline runtime/doc/builtin.txt:183]:
```suggestion
getcmdcompltype()		String	return the type of the current
					command-line completion
```

File: runtime/doc/builtin.txt [hunk 8]
```diff
@@ -2792,6 +2796,13 @@ getcharstr([expr])					*getcharstr()*
 		Otherwise this works like |getchar()|, except that a number
 		result is converted to a string.
 
+getcmdcompltype()					*getcmdcompltype()*
+		Return the current command-line completion type.  Only works
+		when the command line is being edited, thus requires use of
+		|c_CTRL-\_e| or |c_CTRL-R_=|.
```
Comment (zeertzjq) [inline runtime/doc/builtin.txt:2802]:
```suggestion
		Return the type of the current command-line completion.
		Only works when the command line is being edited, thus
		requires use of |c_CTRL-\_e| or |c_CTRL-R_=|.
```

File: runtime/doc/usr_41.txt [hunk 9]
```diff
@@ -855,8 +855,11 @@ Buffers, windows and the argument list:
 	swapname()		get the swap file path of a buffer
 
 Command line:					*command-line-functions*
+	getcmdcompltype()	get the current command line completion
```
Comment (zeertzjq) [inline runtime/doc/usr_41.txt:858]:
```suggestion
	getcmdcompltype()	get the type of the current command line
				completion
```

File: src/nvim/ex_getln.c [hunk 10]
```diff
@@ -5949,6 +5949,23 @@ static struct cmdline_info *get_ccline_ptr(void)
   }
 }
 
+// Get the current command-line completion type.
```
Comment (zeertzjq) [inline src/nvim/ex_getln.c:5952]:
```suggestion
/// Get the current command-line completion type.
```

File: src/nvim/ex_getln.c [hunk 11]
```diff
@@ -5983,6 +6000,16 @@ int get_cmdline_pos(void)
   return p->cmdpos;
 }
 
+int get_cmdline_screen_pos(void)
```
Comment (zeertzjq) [inline src/nvim/ex_getln.c:6003]:
```suggestion
/// Get the command line cursor screen position.
int get_cmdline_screen_pos(void)
```

**Comment by Shougo** [comment]
Thank you the merge!

**Pull request merged**